pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,308 @@
+			sourceNext = sourceIter.Next()
+		case mergeStateEnterDestRange:
+			destNext = destIter.Next()
+		default:
+			// error unknown state
+		}
+		if err != nil {
+			return graveler.DiffSummary{}, err
+		}
+	}
+
+	if sourceNext {
+
+	}
+
+	if destNext {
+
+	}
+
+	if opts != nil && !opts.AllowEmpty && ret.Count[graveler.DiffTypeAdded]+ret.Count[graveler.DiffTypeRemoved]+ret.Count[graveler.DiffTypeChanged] == 0 {
+		return ret, graveler.ErrNoChanges
+	}
+	// handle leftovers
+
+	// only base left -> removed by both (some kind of conflict ) - consider what to do in this case
+	return ret, nil
+}
+
+type compareResult int
+
+const (
+	compareResultRangeConflict compareResult = iota
+	compareResultRangeSourceBefore
+	compareResultRangeSourceBeforeValue
+	compareResultRangeDestBefore
+	compareResultRangeDestBeforeValue
+	compareResultRangeIdentical
+	compareResultValueSourceBefore
+	compareResultValueDestBefore
+	compareResultValueIdentical
+	compareResultValueConflicting
+)
+
+func compareValues(sourceValue *graveler.ValueRecord, destValue *graveler.ValueRecord) compareResult {
+	c := bytes.Compare(sourceValue.Key, destValue.Key)
+	switch {
+	case c < 0:
+		return compareResultValueSourceBefore
+	case c > 0:
+		return compareResultValueDestBefore
+	case bytes.Equal(sourceValue.Identity, destValue.Identity):
+		return compareResultValueIdentical
+	default:
+		return compareResultValueConflicting
+	}
+}
+
+func compare(sourceValue *graveler.ValueRecord, sourceRange *Range, destValue *graveler.ValueRecord, destRange *Range) compareResult {
+	switch {
+	case sourceValue != nil && destValue != nil:
+		return compareValues(sourceValue, destValue)
+	case sourceValue != nil && bytes.Compare(sourceValue.Key, destRange.MinKey) < 0:
+		return compareResultValueSourceBefore
+	case sourceValue != nil && bytes.Compare(destRange.MaxKey, sourceValue.Key) < 0:
+		return compareResultRangeDestBefore
+	case sourceValue != nil:
+		return compareResultRangeDestBeforeValue
+	case destValue != nil && bytes.Compare(destValue.Key, sourceRange.MinKey) < 0:
+		return compareResultValueDestBefore
+	case destValue != nil && bytes.Compare(sourceRange.MaxKey, destValue.Key) < 0:
+		return compareResultRangeSourceBefore
+	case destValue != nil:
+		return compareResultRangeSourceBeforeValue
+	case bytes.Compare(sourceRange.MaxKey, destRange.MinKey) < 0:
+		return compareResultRangeSourceBefore
+	case bytes.Compare(destRange.MaxKey, sourceRange.MinKey) < 0:
+		return compareResultRangeDestBefore
+	case destRange.ID == sourceRange.ID:
+		return compareResultRangeIdentical
+	default:
+		return compareResultRangeConflict
+		// ranges are overlapping with different IDs
+	}
+}
+
+func proceedToRange(iter Iterator, key Key) error {
+	hasMore := true
+	for hasMore {
+		_, rng := iter.Value()
+		if rng == nil {
+			hasMore = iter.Next() // TODO(Guys): change
+		}
+		if rng == nil || bytes.Compare(key, rng.MaxKey) <= 0 {
+			return nil
+		}
+		hasMore = iter.NextRange()
```

Comment:
Doing it this way (compared to seeking) means that we always deserialize every range of base.  That may be expensive.

If you don't want to change the iterator interface, you might instead `iter.SeekGE(key)` to get the range, then seek the iterator to a bit before that range (i.e. subtract 1 from the returned range.MaxKey and pad with lots of 0xFF bytes), and advance to the end of _that_ range.  Still gonna be expensive, but I reckon it'll be a bit faster than churning and deserializing a few megabytes.